### ISSUE TITLE
Rosetta semantic reasoning engine: design checkpoint

### ISSUE THREAD

**Issue by rhythmatician** [rhythmatician/Rosetta#3]
## North star

Rosetta should not become an end-to-end “astrology neural network,” and it should not ask an LLM to reason directly from a pile of chart facts.

The system should become a **question-conditioned semantic reasoning engine over a deterministic astrological graph**:

$$
\boxed{
\text{chart geometry}
\rightarrow
\text{graph dynamics}
\rightarrow
\text{semantic activation field}
\rightarrow
\text{narrative plan}
\rightarrow
\text{LLM narration}
}
$$

The symbolic layer determines **what is structurally present**. The embedding layer represents **what those structures mean**. The LLM handles **communication**, not calculation or unconstrained interpretation.

That is consistent with the product already visible in the repository: deterministic natal, transit, and synastry calculation; pattern detection; chart exploration; and an AI-assisted reading interface.

---

## 1. What the product is actually trying to provide

The ideal product is not a predictor of observable outcomes. There is no defensible target such as income, health, compatibility, or personality score that we are trying to regress from a chart.

The product is closer to:

> Given a deterministically calculated chart and a human question, identify the chart structures most relevant to that question, propagate their meanings through their relationships, and produce an interpretation grounded in both the chart and a curated astrological ontology.

That means the central output of the reasoning engine is not initially prose. It is a structured object containing:

* the active chart entities;
* the relevant aspects and geometric motifs;
* the semantic facets that became active;
* their relative strengths;
* the paths or structures that activated them;
* supporting curated descriptions;
* tensions, reinforcements, missing roles, or ambiguities;
* a narrative outline for the language model.

The LLM receives that object and turns it into readable prose. It should not infer planetary positions, invent chart relationships, choose arbitrary symbolic meanings, or calculate the graph itself.

---

## 2. The conceptual work already present in the repository

The codebase is messy, but her core ideas are recoverable and surprisingly coherent.

### `prompt_templates.py`

This is effectively a long-form product and reasoning specification embedded in Python strings. It contains her rules for grounding, circuit language, qualitative potency, resonance, friction, switch points, membranes, keystones, and staged interpretation.

### `topic_maps.py`

This is a curated mapping from ordinary human concerns to astrological entities:

$$
\text{human topic}\rightarrow
\{\text{planets, signs, houses, points}\}.
$$

It already encodes domains such as identity, relationships, career, creativity, money, emotions, humor, and spirituality, with narrower subtopics and refinements. This is valuable ontology data, not disposable routing code.

### `switch_points.py`

This contains an original product concept:

* incomplete geometric pattern;
* mathematically inferred missing vertex;
* semantic interpretation of that missing role;
* real-world “keystone” that embodies the missing archetype.

It defines transitions such as T-Square → Grand Cross and Wedge → Mystic Rectangle, then connects their geometry to a guided interpretive process.

### `circuit_sim.py`

This is where the linear algebra was already trying to emerge. Planets become power nodes, aspects become weighted edges, patterns become graph topologies, and energy is propagated with hand-authored conductance, resonance, and friction rules. It also uses shortest paths, BFS rerouting, source/sink behavior, and motif-specific parameters.

### `thought_flow.json`

This records the intended reasoning pipeline:

$$
\text{question}
\rightarrow
\text{comprehension}
\rightarrow
\text{routing}
\rightarrow
\text{circuit extraction}
\rightarrow
\text{fact collection}
\rightarrow
\text{reading packet}
\rightarrow
\text{prose}.
$$

It explicitly separates deterministic fact collection from LLM synthesis.

These artifacts should be preserved as source material even if their current implementations are replaced.

---

## 3. The canonical representation

The chart should be represented as a **typed heterogeneous graph**:

$$
G=(V,E,R)
$$

where \(R\) is the set of relation types.

Possible node types include:

* planets and other celestial points;
* chart angles;
* signs;
* houses;
* detected geometric motifs;
* semantic concepts;
* possibly user-question concepts.

Possible edge types include:

* aspect relationships;
* planet-in-sign;
* planet-in-house;
* rulership and dispositorship;
* membership in a detected shape;
* topic-to-astrological-factor mappings;
* concept similarity or opposition;
* switch-point completion relationships.

Some of these may remain attributes rather than first-class nodes. The precise schema remains open, but the important decision is that **everything should not be flattened into one feature vector prematurely**. The graph preserves the structure that gives the chart its meaning.

---

## 4. Where the semantic vectors come from

We should not use the embedding of the bare token `Mars`. That vector contains general-language associations with the word: the Roman god, the planet, candy bars, science fiction, NASA, and many other irrelevant contexts.

We also should not manually extract token-level hidden states and invent an averaging function unless we later decide to train our own encoder. A paragraph embedding model already accepts a complete passage and returns a pooled semantic vector.

For Mars, we would create a **canonical dossier**, but probably not one giant paragraph. It should be a structured collection of passages representing distinct facets, for example:

* astronomical identity;
* basic astrological archetype;
* motivation and agency;
* assertion and conflict;
* desire and pursuit;
* constructive manifestations;
* destructive or dysregulated manifestations;
* behavior under different aspect types;
* boundaries and counterexamples;
* relationships to adjacent concepts such as Sun, Venus, Saturn, and Pluto.

Each passage is embedded separately:

$$
\mathbf z_{M,1},\mathbf z_{M,2},\ldots,\mathbf z_{M,m}
\in\mathbb R^d.
$$

Mars is therefore not necessarily represented by one vector. It can be represented by a matrix:

$$
Z_M=
\begin{bmatrix}
\mathbf z_{M,1}\\
\mathbf z_{M,2}\\
\vdots\\
\mathbf z_{M,m}
\end{bmatrix}.
$$

That lets “Mars as agency” activate in one reading and “Mars as conflict” activate in another.

### A representative vector, when needed

A simple representative vector could be a normalized weighted centroid:

$$
\boldsymbol{\mu}_M
=
\frac{\sum_i w_i\mathbf z_{M,i}}
{\left\|\sum_i w_i\mathbf z_{M,i}\right\|}.
$$

But we should retain the individual facet vectors rather than throwing them away. The centroid is useful for broad retrieval; the facets retain nuance.

A distributional view could also retain:

$$
\boldsymbol{\mu}_M
\quad\text{and}\quad
\Sigma_M,
$$

where \(\Sigma_M\) describes dispersion among Mars passages. High dispersion would indicate that the dossier contains several distinct senses or poorly separated facets.

In practice, explicit facet labels and clusters will probably be more stable and interpretable than trying to estimate a full covariance matrix from a small number of passages.

---

## 5. Language embeddings plus the curated graph

The strongest representation will probably combine two sources:

1. **Language geometry**, obtained from embedding the curated descriptions.
2. **Domain geometry**, obtained from her explicit ontology and graph relationships.

The base embedding model knows that “assertion,” “initiative,” and “action” are semantically related. Her ontology knows that those concepts are linked to Mars, Aries, the first house, certain aspects, and particular chart structures.

A mathematically clean way to combine them is graph-based embedding retrofitting.

Let:

* \(B\in\mathbb R^{n\times d}\) contain the original language embeddings;
* \(L\) be the graph Laplacian of the curated concept graph;
* \(Z\) be the adjusted domain embeddings.

We could optimize:

$$
Z^*
=
\arg\min_Z
\left[
\|Z-B\|_F^2
+
\lambda\operatorname{tr}(Z^\mathsf{T}LZ)
\right].
$$

The first term prevents the vectors from drifting too far from their language meanings. The second pulls concepts together when the curated ontology says they are related.

In the simplest form, the solution is:

$$
\boxed{
Z^*=(I+\lambda L)^{-1}B
}
$$

or an iterative equivalent that avoids explicitly inverting a large matrix.

This is not yet a settled implementation decision, but it is a strong candidate for the elegant mathematical bridge between the pretrained semantic space and her handcrafted ontology.

Counterexamples and explicit oppositions could later add margin constraints so that unrelated or contrasting concepts are not collapsed together.

---

## 6. How meaning propagates through the chart

Let \(X\in\mathbb R^{n\times d}\) be the semantic feature matrix for the active chart nodes. Each row is a node or facet representation.

Let \(A_r\) be a normalized adjacency matrix for relation type \(r\), such as:

* trine;
* square;
* opposition;
* conjunction;
* rulership;
* motif membership.

A simple interpretable propagation model could be:

$$
H^{(0)}=X
$$

$$
H^{(k+1)}
=
\alpha H^{(0)}
+
\sum_r \beta_r A_r H^{(k)}.
$$

Here:

* \(\alpha\) retains each node’s original identity;
* \(\beta_r\) controls how much meaning propagates across relation \(r\);
* repeated steps allow influence to travel through larger chart structures.

A nonlinearity, clipping, normalization, or gating mechanism could be added later, but the first version does not need to be a neural network.

A finite graph-diffusion form would be:

$$
H
=
\alpha_0X
+\alpha_1AX
+\alpha_2A^2X
+\cdots+\alpha_kA^kX.
$$

That is the linear algebra hiding under the current circuit simulation. Her present code manually computes local power transmission, friction, rerouting, source/sink behavior, and topology effects. The redesigned version would express those operations with a smaller number of explicit matrix operations and typed relation coefficients.

---

## 7. Question-conditioned interpretation

A chart does not have one final meaning vector. The relevant semantic state depends on the user’s question.

Let \(\mathbf q\in\mathbb R^d\) be the embedding of the structured question or question concept.

We can compute attention-like relevance over graph states:

$$
a_i
=
\operatorname{softmax}
\left(
\tau\,\mathbf q^\mathsf{T}\mathbf h_i
+p_i
\right),
$$

where \(p_i\) can incorporate deterministic relevance from the topic map, chart structure, or focus selection.

The question-conditioned chart representation is then:

$$
\mathbf s
=
\sum_i a_i\mathbf h_i.
$$

Relevant concepts are retrieved by similarity:

$$
\operatorname{score}(c_j)
=
\cos(\mathbf s,\mathbf e_{c_j}).
$$

The output should not merely list the nearest concepts. It should preserve provenance:

```text
Concept: disciplined assertion
Activated by:
  Mars facet: agency
  Saturn facet: constraint/structure
  Mars square Saturn
  relevant to user question: conflict between action and responsibility
Contribution:
  direct Mars relevance
  square-mediated Saturn influence
  topic-map support
```

That provenance is what makes the engine debuggable, testable, and suitable for reliable LLM narration.

---

## 8 What happens to the arbitrary model weights

The degree boundaries, signs, aspect definitions, orbs, and shape-detection rules are largely part of the astrological formal system. Those should be treated separately from the model’s interpretive weights.

The quantities that require rethinking are things like:

* aspect conductance;
* topology resonance;
* topology friction;
* source/sink bonuses;
* mutual-reception bonuses;
* motif-strength multipliers;
* propagation depth;
* relative contributions of node identity, edges, and motifs.

The current code treats values such as `0.95 resonance` and `0.05 friction` as table entries. We agreed that these numbers probably should not all remain independent.

### First reduce the parameterization

Rather than independently assigning two values to every topology, derive them from structural features such as:

* edge-type composition;
* cycle closure;
* path redundancy;
* symmetry;
* algebraic connectivity;
* balance around the zodiac;
* motif completeness;
* number and arrangement of oppositions;
* node potency distribution;
* source-to-sink connectivity.

For topology \(t\), define a feature vector:

$$
\boldsymbol{\phi}_t.
$$

Then resonance could be:

$$
r_t
=
\sigma
\left(
\beta_0+\boldsymbol{\beta}^\mathsf{T}\boldsymbol{\phi}_t
\right).
$$

Friction could be modeled separately, or resonance/friction/neutral behavior could be constrained with a softmax:

$$
(r_t,f_t,n_t)
=
\operatorname{softmax}(W\boldsymbol{\phi}_t).
$$

This prevents dozens of unrelated magic numbers from accumulating.

### Preserve qualitative ordering

Her intended relationships may be more reliable than her exact numbers. For example, she may believe one topology is more resonant than another without knowing whether the values should be 0.80 and 0.70 or 0.65 and 0.45.

We can encode pairwise constraints:

$$
r_{\text{Grand Trine}}
>
r_{\text{T-Square}}.
$$

That yields ranking losses rather than demanding arbitrary numeric targets.

### Do not assume conservation without defining it

Resonance and friction currently often look complementary, but we should not automatically assert:

$$
r_t+f_t=1.
$$

That is a modeling choice, not a discovered law. There may be structures that are simultaneously highly resonant and highly tense. A three-way or multi-axis representation may describe the ontology better than one slider from harmony to friction.

---

## 9. How the weights can be fitted without a real-world regressor

We do not have a target variable describing whether astrology is objectively correct. We do, however, have several possible forms of supervision.

### A. Fit to her curated interpretations

Her descriptions define what semantic content she expects entities, aspects, and patterns to produce.

For training example \(k\):

* the graph model produces semantic state \(\hat{\mathbf y}_k(\theta)\);
* the curated interpretation is embedded as \(\mathbf y_k\);
* \(\theta\) contains the small set of propagation and structural parameters.

Optimize:

$$
\mathcal L_{\text{semantic}}
=
\sum_k
\left[
1-
\cos
\left(
\hat{\mathbf y}_k(\theta),
\mathbf y_k
\right)
\right].
$$

This does not claim the interpretation is empirically true. It fits the engine to **her intended semantic system**.

### B. Fit to concept sets from the topic maps

Where prose is unavailable, mappings such as:

$$
\text{self-confidence}
\rightarrow
\{\text{Sun, Mars, Mercury, Venus, Leo, 5th House}\}
$$

provide weak supervision.

The model can be rewarded when a question embedding for “self-confidence” activates those entities and their appropriate facets more strongly than unrelated ones.

### C. Fit qualitative rankings

She can provide cheap judgments such as:

* interpretation A is closer than B;
* this concept should rank above that concept;
* this shape should be more resonant than that one;
* this aspect should propagate tension more strongly than support.

Pairwise comparison is faster and more reliable than asking her to assign numeric scores.

### D. Fit internal consistency

The same semantic relationship may appear in topic maps, profile text, prompt instructions, and shape descriptions. Parameters can be chosen to minimize disagreement among those representations.

### E. Use later user feedback only as secondary calibration

If the application eventually accumulates reading ratings or pairwise preferences, those can calibrate the system. They should not be required for the first useful model.

A combined objective might be:

$$
\mathcal L(\theta)
=
\mathcal L_{\text{semantic}}
+
\lambda_r\mathcal L_{\text{ranking}}
+
\lambda_c\mathcal L_{\text{consistency}}
+
\lambda_2\|\theta\|_2^2.
$$

This is a genuine constrained optimization or regression problem. Its target is fidelity to the curated ontology, not measurable life outcomes.

---

## 10. What we are not planning to do initially

We are not planning to:

* train an end-to-end neural network to predict readings;
* fine-tune a large language model;
* treat a vector database as the reasoning engine;
* manually average raw LLM token embeddings;
* reduce the semantic space to the number of planets;
* let an LLM decide which chart facts exist;
* preserve the current arbitrary numeric table as astrological truth;
* judge every parameter change by generating and reading a complete essay.

A vector database may eventually store and retrieve dossiers, facets, concepts, and supporting passages. It is infrastructure, not the semantic model itself.

A graph neural network could become useful after the system has accumulated a substantial set of chart–question–interpretation examples. It is not necessary to build the first principled version.

---

## 11. The shortened evaluation loop

Most testing should stop before prose generation.

For a small chart fixture and question, the harness should show:

```text
Relevant chart structures
1. Mars square Saturn
2. T-Square containing Mars, Saturn, Moon

Activated semantic facets
1. constrained agency
2. frustration under limitation
3. disciplined effort
4. conflict between urgency and structure

Provenance
- Mars/agency direct activation
- Saturn/constraint propagated through square
- T-Square pressure-funnel motif
- question matched career/leadership topic map

Suppressed or competing facets
- spontaneity
- ease
- passive receptivity
```

Her evaluation becomes:

* Is the correct structure active?
* Are the top concepts appropriate?
* Is anything important missing?
* Is something irrelevant dominating?
* Does the provenance make sense?
* Are relative rankings correct?

Full readings become periodic integration tests rather than the inner development loop.

---

## 12. Architecture that should make the code easier for agents

After cleanup, the system should have a small number of authoritative layers:

```text
chart/
    deterministic astronomy and astrology calculations

ontology/
    entities, facets, topic maps, curated descriptions, relationships

graph/
    typed chart graph construction and motif detection

semantics/
    embedding generation, storage, retrofitting, similarity

reasoning/
    propagation, question conditioning, activation ranking, provenance

narrative/
    deterministic narrative plan and LLM prompt assembly

ui/
    interaction and visualization
```

Long ontology prose should move out of Python strings and into versioned, validated data files. Each concept should have a stable identifier, metadata, source, facets, relationships, and revision history.

The mathematical core should consist primarily of pure functions over typed data structures. That gives agents much smaller edit surfaces and makes tests possible without launching the UI or generating prose.

---

### Settled

* Deterministic chart computation remains authoritative.
* The chart becomes a typed graph.
* Embeddings are foundational semantic representations, not an optional retrieval feature added at the end.
* Entities should probably have multiple facet vectors rather than one monolithic vector.
* A paragraph embedding model creates those vectors directly.
* Graph operations propagate and combine semantics.
* The user’s question conditions the semantic readout.
* The LLM narrates a structured, provenance-bearing result.
* Arbitrary weight tables should be reduced to a smaller constrained parameterization.
* Her curated corpus can provide semantic supervision without claiming empirical astrology validation.
* The inner test loop should evaluate concept activation and provenance, not complete readings.

### Strong working hypotheses

* Combine language embeddings with the curated ontology through graph retrofitting or a related Laplacian method.
* Derive motif weights from structural graph features rather than topology names alone.
* Fit a small number of propagation coefficients by semantic alignment and ranking constraints.
* Keep resonance and friction as potentially distinct dimensions rather than forcing them to sum to one.
* Treat existing prose and topic mappings as a dataset to preserve, normalize, and evaluate against.

### Still open

* The exact graph schema and which objects deserve nodes.
* The embedding model.
* The canonical facet taxonomy.
* How much astronomical description belongs in the same space as symbolic astrological meaning.
* Which existing interpretations are genuinely hers, imported from other sources, or generated by prior agents.
* The initial training/evaluation examples.
* The structural features used to parameterize topology behavior.
* Whether graph retrofitting is sufficient or contrastive learning is needed.
* The appropriate number of propagation steps.
* How aggressively to preserve versus revise her current circuit metaphors.
* The product boundary between explanation, advice, and prediction.

---

## Starting point for tomorrow

The next conceptual task is to define one complete vertical slice using **Mars** as the test entity:

1. Specify the canonical Mars dossier and its facets.
2. Embed each facet.
3. Connect Mars to relevant human concepts from the existing ontology.
4. Define how one aspect, probably Mars square Saturn, transforms or combines those facets.
5. Run a question-conditioned readout for a few small questions.
6. Inspect the activated concepts and provenance without generating prose.
7. Use that experiment to determine the canonical graph schema and the first weight-fitting objective.

That should tell us whether the architecture works before turning it into a repository-wide specification.
